Add SARIF report output

yongyan-gh opened the pull request:
For internal review.
Known issue:
- Missing Rule name.

Attached sample SARIF report file: [result.zip](https://github.com/yongyan-gh/template-analyzer/files/7445486/result.zip)
Please review.

yongyan-gh committed:
Add SARIF report output

eddynaka commented on src/Analyzer.Cli/CommandLineParser.cs:
```diff
@@ -71,7 +81,11 @@ private RootCommand SetupCommandLineAPI()
 
-            analyzeDirectoryCommand.Handler = CommandHandler.Create<DirectoryInfo>((directoryPath) => this.AnalyzeDirectory(directoryPath));
+            analyzeDirectoryCommand.AddOption(reportFormatOption);
+
```
remove this new line #Resolved

eddynaka commented on src/Analyzer.Cli/CommandLineParser.cs:
```diff
@@ -91,6 +105,13 @@ private int AnalyzeTemplate(FileInfo templateFilePath, FileInfo parametersFilePa
                 }
 
+                // Check that output file path provided for sarif report
+                if (writer == null && reportFormat != ReportFormat.Console && outputFilePath == null)
```
Check for the specific format. Imagine in the future, you have a different format. #Resolved

eddynaka commented on src/Analyzer.Cli/CommandLineParser.cs:
```diff
@@ -91,6 +105,13 @@ private int AnalyzeTemplate(FileInfo templateFilePath, FileInfo parametersFilePa
+                // Check that output file path provided for sarif report
+                if (writer == null && reportFormat != ReportFormat.Console && outputFilePath == null)
+                {
+                    Console.WriteLine($"Output file path is not provided.");
```
```suggestion
                    Console.WriteLine($"Output file path was not provided.");
``` #Resolved

eddynaka commented on src/Analyzer.Cli/CommandLineParser.cs:
```diff
@@ -80,7 +94,7 @@ private RootCommand SetupCommandLineAPI()
         }
 
-        private int AnalyzeTemplate(FileInfo templateFilePath, FileInfo parametersFilePath, bool printMessageIfNotTemplate = true)
+        private int AnalyzeTemplate(FileInfo templateFilePath, FileInfo parametersFilePath, ReportFormat reportFormat, FileInfo outputFilePath, bool printMessageIfNotTemplate = true, IReportWriter writer = null)
```
what is the case when writer is null? Per my understanding, you did create two writers Console/Sarif

yongyan-gh replied:
the tool can analyze a directory or a template file.
the existing function for analyze directory is to enumerate all template files under the directory then call AnalyzeTemplate for each template file.

If tool analyzes a directory, the writer is initialized at AnalyzeDirectory function, and pass the write to AnalyzeTemplate function so it will not initialize a new writer again.

If tool analyzes a template file, it call AnalyzeTemplate with null writer, so the function will create its own writer in this case.

eddynaka commented on src/Analyzer.Reports/ReportFormat.cs:
```diff
@@ -0,0 +1,22 @@
+        Console = 0,
+
+        /// <summary>
+        /// output to sarif report
```
```suggestion
        /// Output to SARIF report.
``` #Resolved

eddynaka commented on src/Analyzer.Reports/ReportFormat.cs:
```diff
@@ -0,0 +1,22 @@
+    public enum ReportFormat
+    {
+        /// <summary>
+        /// output to console
```
```suggestion
        /// Output to console.
``` #Resolved

eddynaka commented on src/Analyzer.Reports/SarifReportWriter.cs:
```diff
@@ -0,0 +1,188 @@
+        // may define the const values in common place
+        private const string ToolName = "ARM BPA";
+        private const string ToolFullName = "ARM Template Best Practice Analyzer";
+        private const string ToolVersion = "0.0.2-alpha";
```
This should not be hard-coded.

yongyan-gh replied:
Agree, asked the team there is no official version defined in code yet. Will double check with them. Will also add a comment here.

eddynaka commented on src/Analyzer.Reports/SarifReportWriter.cs:
```diff
@@ -0,0 +1,188 @@
+
+        private FailureLevel GetLevelFromEvaluation(IEvaluation evaluation)
+        {
+            //rule severity in designing https://github.com/Azure/template-analyzer/issues/177
```
```suggestion
            // The rule severity definition: https://github.com/Azure/template-analyzer/issues/177
``` #Resolved

eddynaka commented on src/Analyzer.Reports/SarifReportWriter.cs:
```diff
@@ -0,0 +1,188 @@
+            using (FileStream outputTextStream = this.reportFile.Create())
+            using (var outputTextWriter = new StreamWriter(outputTextStream))
+            using (var outputJsonWriter = new JsonTextWriter(outputTextWriter) { Formatting = Formatting.Indented })
+            using (var output = new ResultLogJsonWriter(outputJsonWriter))
```
You can replace almost everything for the SarifLogger.

yongyan-gh replied:
updated pls review

yongyan-gh commented:
cannot add any people as reviewer in this forked repo. so adding here @eddynaka @shaopeng-gh @michaelcfanning.
please review. unit tests will be added soon.

yongyan-gh committed:
Add Analyzer.Reports.UnitTests

Throw exceptions related to incorrect segment lengths

Add test for incorrect segment lengths exception

Change TA-000028 so that its test resources can define correct segment lengths

Improve TA-000028 to temporarily support both ways of defining resource names and types in a nested resource

Fix regex to match resource names that end in /default or are only default and add test cases for it

Do not re-throw exceptions about incorrect segment lengths

Add unit tests & E2E tests

yongyan-gh committed:
Add period to sarif rule message text
Remove redundant sarif result message text
Add console report writer tests

yongyan-gh committed:
Merge branch 'main' into users/yongyan-gh/SarifOutput

yongyan-gh committed:
Using SarifLogger instead
Fix test cases

yongyan-gh committed:
merge from main branch

yongyan-gh committed:
Fix build error & address review feedbacks

yongyan-gh committed:
Merge branch 'main' into users/yongyan-gh/SarifOutput

yongyan-gh committed:
Refactor unit tests & address review feedbacks

yongyan-gh committed:
Merge branch 'users/yongyan-gh/SarifOutput' of https://github.com/yongyan-gh/template-analyzer into users/yongyan-gh/SarifOutput

yongyan-gh committed:
Address feedbacks

yongyan-gh committed:
Refactor SarifReportWriter to persist results per file analyzed.

yongyan-gh committed:
fix null check

yongyan-gh committed:
Apply suggestions from code review

Co-authored-by: Vera Bogdanich Espina <VeraBE@users.noreply.github.com>

JohnathonMohr committed:
Merge branch 'main' into users/yongyan-gh/SarifOutput

yongyan-gh committed:
Update tests to use portal file paths.

yongyan-gh committed:
Merge branch 'users/yongyan-gh/SarifOutput' of https://github.com/yongyan-gh/template-analyzer into users/yongyan-gh/SarifOutput

yongyan-gh committed:
Fix test failures due to relative path
